pkg/stanza/fileconsumer/reader.go

```diff
@@ -58,12 +60,34 @@ func (r *Reader) offsetToEnd() error {
 
 // ReadToEnd will read until the end of the file
 func (r *Reader) ReadToEnd(ctx context.Context) {
-	if _, err := r.file.Seek(r.Offset, 0); err != nil {
+	// read through the fingerprintReader in order to update the fingerprint as we read.
+	fpr := &fingerprintReader{
+		offset:          r.Offset,
+		fingerprintSize: r.fingerprintSize,
+		file:            r.file,
+		fingerprint:     r.Fingerprint,
+	}
+
+	if r.header != nil && !r.header.Finalized() {
+		r.header.ReadHeader(ctx, fpr, r.maxLogSize, r.encoding, r.fileAttributes)
+		// Don't read log entries if the header has not yet been finalized
+		// (we are still waiting for the full header to be read).
+		if !r.header.Finalized() {
+			return
+		}
+
+		// Set r to the end of the headers if our current offset is within the header
+		if r.Offset < r.header.Offset() {
+			r.Offset = r.header.Offset()
+		}
+	}
+
+	if _, err := fpr.Seek(r.Offset, 0); err != nil {
```

Comment:
Do we really need to duplicate or rework any of the code relating to how the lines are read? What if we just change how the lines are processed?

```
type Reader struct {
	...
	processFunc EmitFunc
}

// when building reader
if r.header != nil {
	r.processFunc = r.header. processHeaderLine
}

// when header has been fully read
r.processFunc = r.emit
```

Reply:
I had tried this out originally, but here's what got me to change it: 
1. Offsets (could probably be bypassed by implementing persistence)
2. Starting to read from the end (we could just say this is not supported)
3. Multiline split functions for log lines might break the header parsing (header lines are typically formatted different from the actual log lines, and the multiline config for log lines is likely not compatible with the header lines).

It's certainly simpler to do it that way, but it has some limitations that make it difficult in other areas.

Reply:
> I had tried this out originally, but here's what got me to change it:
> 
> 1. Offsets (might be bypassed by implementing persistence?)
> 2. Starting to read from the end (we could just say this is not supported)
> 3. Multiline split functions for log lines might break the header parsing (header lines are typically formatted different from the actual log lines, and the multiline config for log lines is likely not compatible with the header lines).
> 
> It's certainly simpler to do it that way, but it has some limitations that make it difficult in other areas.

Might be worth it:

1. If we are able to implement header processing within the existing loop, we may already be persisting everything we need. I'm curious if there's anything in particular you think would need to change here.
2. See point above
3. Maybe we can rebuild/update the scanner with a new splitfunc once we are through the header

Reply:
Tried reworking it like this;

It does seem like it would add more complexity to the read loop to implement starting to read from the end of the file (`start_at` = `end`), so I've dropped support of that for now.